syntax = "proto3";

import "metadata.proto";

package loot.protobuf;

message LootSettings {
    bool enable_debug_logging = 1;
    bool update_masterlist = 2;
    string game = 3;
    string last_game = 4;
    string last_version = 5;
    string language = 6;
    WindowPosition window_position = 7;
    repeated GameSettings games = 8;
    map<string, bool> filters = 9;

    message WindowPosition {
        int64 top = 1;
        int64 bottom = 2;
        int64 left = 3;
        int64 right = 4;
        bool maximised = 5;
    }

    message GameSettings {
        string type = 1;
        string name = 2;
        string master = 3;
        string registry = 4;
        string folder = 5;
        string repo = 6;
        string branch = 7;
        string path = 8;
    }
}

message SimpleMessage {
    string type = 1;
    //Type type = 1;
    string text = 2;
    string language = 3;
    string condition = 4;

    // Not currently used as types are strings.
    enum Type {
        SAY = 0;
        WARN = 1;
        ERROR = 2;
    }
}

message PluginMetadata {
    string name = 1;
    bool enabled = 2;
    sint32 priority = 3;
    sint32 global_priority = 4;
    repeated File after = 5;
    repeated File req = 6;
    repeated File inc = 7;
    repeated SimpleMessage msg = 8;
    repeated Tag tag = 9;
    repeated CleaningData dirty = 10;
    repeated CleaningData clean = 11;
    repeated Location url = 12;
}

message Plugin {
    string name = 1;
    string version = 2;
    bool is_active = 3;
    bool is_empty = 4;
    bool is_master = 5;
    bool loads_archive = 6;

    uint32 crc = 7;
    sint32 load_order_index = 8;

    sint32 priority = 9;
    sint32 global_priority = 10;
    bool is_dirty = 11;
    string cleaned_with = 12;
    repeated SimpleMessage messages = 13;
    repeated Tag tags = 14;

    PluginMetadata masterlist = 15;
    PluginMetadata userlist = 16;
}
